source/Agents/TrackingAgent.hpp

```diff
@@ -0,0 +1,512 @@
+/**
+ * This file is part of the Fall 2023, CSE 491 Course Project
+ * @file TrackingAgent.hpp
+ * @brief Agent that switches between user-defined custom movement pattern and
+ * tracking a given agent
+ * @note Status: PROPOSAL
+ * @author David Rackerby
+ * @author Matt Kight
+ **/
+
+#pragma once
+
+#include "../core/AgentBase.hpp"
+#include "../core/GridPosition.hpp"
+#include "AStarAgent.hpp"
+#include "PathAgent.hpp"
+#include "AgentLibary.hpp"
+
+#include <memory>
+#include <limits>
+#include <string>
+#include <string_view>
+#include <type_traits>
+#include <unordered_set>
+#include <variant>
+#include <vector>
+
+namespace walle {
+
+    /**
+     * A single Alerter is responsible for forcefully changing the state of all trackers in its network
+     * to TrackingState::TRACKING when a single TrackingAgent in the set of trackers
+     * comes into range of its goal_pos
+     */
+    class Alerter {
+    private:
+        /// World the agents in the agent network are a part of
+        cse491::WorldBase *world_ptr_;
+
+        /// Set of all agents in a single network
+        std::unordered_set<size_t> agent_network_set_;
+
+    public:
+        // Must be constructed with an associated world
+        Alerter() = delete;
+        explicit Alerter(cse491::WorldBase *world_ptr);
+        Alerter(cse491::WorldBase *world_ptr, size_t id);
+        void AddAgent(size_t id);
+        void RemoveAgent(size_t id);
+        void AlertAllTrackingAgents(size_t caller_id);
+
+        /// Returns an immutable reference to the Alerter's set of agents it knows about
+        std::unordered_set<size_t> const& GetNetworkSet() const { return agent_network_set_; }
+    }; // Alerter
+
+        /**
+         * Used to keep track of what action we are currently taking
+         */
+        enum class TrackingState { RETURNING_TO_START, TRACKING, PATROLLING };
```

Comment:
Looks like indentation is off from here down. 
Not a huge deal, but a little confusing at a glance.

Reply:
Ah, only until the TrackingAgent class a few lines down.